### PR TITLE
Add HMAC request stamping for REST API auth

### PR DESCRIPTION
**Pull request by ohitslaurence**
## Summary
- Add `integratorSecret` client option and HMAC-SHA256 request stamper for REST API calls
- REST API requests are signed with `X-Integrator-Key`, `X-Signature`, and `X-Timestamp` headers matching the platform's integrator HMAC auth spec
- Add query parameter support to `restApi()` with canonical sorting for signature stability
- Uses Web Crypto API (`crypto.subtle`) — works in Node 20+ and browsers, no new dependencies

## Test plan
- [x] 25 new tests covering canonicalization, signature generation, stamping, and cross-compat
- [x] All existing tests pass
- [x] Build succeeds

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lib/client.ts [hunk 1]
```diff
@@ -49,20 +50,24 @@ export class SpritzClient {
...
         environment,
         timeout = 300000, // 5 minutes
         apiKey,
         integrationKey,
+        integratorSecret,
     }: {
         environment: Environment
         timeout?: number | undefined
         integrationKey: string | undefined
+        integratorSecret: string | undefined
```
Comment (ohitslaurence) [inline src/lib/client.ts:65]: **[Clarity · Suggestion]** `integratorSecret` is required in `SpritzClient` constructor type but optional in `ClientOptions`

The `SpritzClient` constructor inline type declares `integratorSecret: string | undefined` (required parameter that accepts undefined), while `ClientOptions` declares it as `integratorSecret?: string` (truly optional). This works at runtime because `SpritzApiClient.init()` always passes it through, but the `SpritzClient` constructor type is misleading -- it signals that callers *must* think about this parameter even when they have no integrator secret. Consider `integratorSecret?: string | undefined` (optional) to match the intent.

Same issue existed for `integrationKey` and `apiKey` before this PR, so this is inherited awkwardness. But since you are already touching this constructor signature, reasonable time to fix the pattern for the new param.

File: src/lib/hmac.test.ts [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        it('produces correct payload structure for GET without body', async () => {
+            const timestamp = 1700000000000
+            const sig = await generateHmacSignature(testSecret, timestamp, {
+                method: 'GET',
+                path: '/v1/users/me',
+            })
+            // Empty body hash should be empty string in payload
+            expect(sig).toMatch(/^sha256=[0-9a-f]{64}$/)
+        })
+    })
```
Comment (ohitslaurence) [inline src/lib/hmac.test.ts:304]: **[Test · Concern]** Cross-compatibility tests don't verify cross-compatibility

These two tests are labeled "cross-compatibility with platform server" but they only check that (a) the same inputs produce the same output (already tested in `generateHmacSignature` > "generates consistent signatures") and (b) the output matches the `sha256=[hex]{64}` format (also already tested). Neither test asserts against a known-good signature value from the platform server.

To actually verify wire compatibility, compute the expected signature outside this code (e.g., with the platform's Node `crypto.createHmac` implementation or a CLI tool) for a fixed secret/timestamp/method/path/body, then assert the exact hex digest here. Without that, a payload-construction bug (wrong field order, wrong separator, wrong body-hash behavior) would pass all tests while producing signatures the server rejects.

File: src/lib/hmac.test.ts [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        it('produces consistent output regardless of param order', () => {
+            const url1 = new URL(
+                'https://api.example.com/v1/on-ramps?network=ethereum&token=USDC&limit=20',
+            )
+            const url2 = new URL(
+                'https://api.example.com/v1/on-ramps?limit=20&token=USDC&network=ethereum',
+            )
+            expect(canonicalizeQueryString(url1)).toBe(canonicalizeQueryString(url2))
+        })
+    })
```
Comment (ohitslaurence) [inline src/lib/hmac.test.ts:56]: **[Test · Concern]** Duplicate query param keys silently produce wrong signatures — untested

`canonicalizeQueryString` iterates `params.keys()` which yields duplicate entries for repeated keys (e.g., `?tag=a&tag=b` yields `tag, tag`), but then calls `params.get(key)` which only returns the *first* value. The result is `tag=a&tag=a` — the second value is silently dropped, producing a different signature than the server would if it preserves all values.

This is both a code bug and a test gap. If the platform server uses `getAll()` or iterates entries, signatures will diverge on any multi-valued param. Add a test with duplicate keys to surface this, and consider whether the implementation should use `params.entries()` or `[...params]` instead of `keys()` + `get()`.

File: src/lib/hmac.test.ts [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                path: '/v1/transactions',
+                body: '{"amount":100}',
+            })
+            const sig2 = await generateHmacSignature(testSecret, timestamp, {
+                method: 'POST',
+                path: '/v1/transactions',
+                body: '{"amount":100}',
+            })
+            expect(sig1).toBe(sig2)
+        })
```
Comment (ohitslaurence) [inline src/lib/hmac.test.ts:280]: **[Clarity · Suggestion]** Cross-compat tests do not actually cross-verify against the platform

The "cross-compatibility with platform server" describe block comments claim to verify parity with the platform's Node crypto implementation, but the tests only check that this implementation is deterministic (same inputs → same output) and that the output format matches `sha256=[hex]`. They do not assert against any known-good signatures computed by the platform server.

Consider adding at least one golden-value test: compute the expected signature with the platform's implementation for a fixed (secret, timestamp, method, path, body) tuple, then assert this implementation matches. Without that, these tests do not actually verify cross-compatibility -- they only verify internal consistency, which is already covered by the `generateHmacSignature` tests above.

File: src/lib/hmac.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    const params = url.searchParams
+    if ([...params].length === 0) return ''
+
+    return [...params.keys()]
+        .sort()
+        .map((key) => {
+            const value = params.get(key) ?? ''
+            return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
+        })
+        .join('&')
```
Comment (ohitslaurence) [inline src/lib/hmac.ts:42]: **[Craft · Concern]** `canonicalizeQueryString` silently corrupts duplicate query parameter keys

`[...params.keys()]` yields duplicate entries for multi-value params (e.g., `?tag=a&tag=b`), but `params.get(key)` always returns only the *first* value. The result would be `tag=a&tag=a` — a silent data corruption that produces a valid-looking but wrong signature.

The current calling code uses `searchParams.set()` which prevents duplicates, but `canonicalizeQueryString` is an exported function that accepts an arbitrary `URL`. It should be correct on its own. Using `[...params.entries()]` fixes this and simplifies the code:

```ts
return [...url.searchParams.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
    .join('&')
```

File: src/lib/hmac.test.ts (hunk 2 above)
Comment (ohitslaurence) [inline src/lib/hmac.test.ts:304]: **[Craft · Suggestion]** Cross-compatibility tests don't actually verify cross-compatibility

These tests only confirm determinism (same input produces same output), which is already covered by the `generateHmacSignature` tests above. To verify the client matches the platform's implementation, you need at least one known test vector — a fixed secret, timestamp, method, path, and body with the *expected* signature output from the server. Without that, the describe label overpromises.

Either add a pinned test vector from the platform (preferred) or rename this block to something like "payload construction determinism."

**Comment by ohitslaurence** [comment]
## PR Clarity & Completeness Review

**Description accuracy**: The PR description is accurate and complete. All four bullet points match the actual diff. No undocumented changes.

**Scope**: This is one cohesive logical change. The query param support in `restApi()` is tightly coupled to the HMAC work (query params must be canonicalized for signature stability), so bundling it here is appropriate rather than a drive-by change.

**Intent clarity**: Good. The description explains *what* and *how* well (Web Crypto, header names, no new deps). A reviewer unfamiliar with the codebase could understand the purpose. One minor gap: the description does not explain *when* stamping is applied -- it is only applied to `sendRestApiRequest`, not to `sendRequest` (the older REST path) or GraphQL queries. This is presumably intentional but worth a one-liner in the description for reviewers who might wonder.

**Loose ends**: No TODO/FIXME/HACK markers introduced.

**API surface / breaking changes**: This is a purely additive, non-breaking change. `integratorSecret` is optional in `ClientOptions`. The new `query` param on `restApi()` is optional. No migration needed. The `SpritzApiClient` constructor (which is not part of the public API -- consumers use `initialize()`) gained a positional parameter, but since it is optional and appended at the end, existing internal callers are unaffected.

**Missing changeset**: This PR adds a new public client option (`integratorSecret`) and new request-signing behavior. It should have a changeset (minor bump) so the change appears in the changelog when published. Currently there is no `.changeset/` entry in the diff.

File: src/lib/client.ts [hunk 6]
```diff
@@ -160,12 +168,26 @@ export class SpritzClient {
...
     }: {
         method: HTTPMethod
         path: string
         body: GraphQLVariables | undefined
+        query?: Record<string, string | number | boolean | undefined>
     }) {
-        const { url, req, timeout } = this.buildRestRequest(method, path, body, this.baseRestApiURL)
+        const { url, req, timeout } = this.buildRestRequest(method, path, body, this.baseRestApiURL, query)
+
+        if (this.integrationKey && this.integratorSecret) {
```
Comment (ohitslaurence) [inline src/lib/client.ts:180]: **[Craft · Suggestion]** `integratorSecret` without `integrationKey` is silently ignored

If someone provides `integratorSecret` but forgets `integrationKey`, requests go out unsigned with no indication anything is wrong. This is the kind of misconfiguration that's painful to debug in production — everything appears to work until the server rejects the request.

Consider either validating at construction time that `integratorSecret` requires `integrationKey` (throw early), or at minimum logging a warning here when only one of the pair is present.

Comment (ohitslaurence) [inline src/lib/client.ts:180]: **[Arch · Concern]** Silent no-op when `integratorSecret` is provided without `integrationKey` (or vice versa)

The guard `if (this.integrationKey && this.integratorSecret)` silently skips signing when only one of the two is provided. For a security-critical auth mechanism, a misconfigured client should fail loudly, not send unsigned requests that the server will reject with an opaque 401. Consider validating at construction time (in `SpritzApiClient` or `SpritzClient`) that these two options are either both present or both absent, similar to the existing `apiKey`/`integrationKey` check.

**Comment by ohitslaurence** [comment]
## PR Review

| Dimension | Concerns | Suggestions | Nits |
|-----------|----------|-------------|------|
| Architectural Fit | 1 | 0 | 0 |
| Craft & Readability | 1 | 1 | 0 |
| Testing Integrity | 2 | 0 | 0 |
| PR Clarity | 0 | 2 | 0 |

**Overall**: The HMAC stamper is well-structured and placed correctly. Three findings worth addressing before merge:

1. **Duplicate query param keys** — `canonicalizeQueryString` uses `keys()` + `get()` which silently drops repeated values (`?tag=a&tag=b` → `tag=a&tag=a`). Use `entries()` instead.
2. **Silent misconfiguration** — providing `integratorSecret` without `integrationKey` (or vice versa) silently skips signing. Should fail at construction time.
3. **Cross-compat tests are mislabeled** — they only verify determinism, not actual parity with the platform. Add a golden test vector from the server-side implementation.

Also: missing changeset for the new public `integratorSecret` option.

File: src/lib/hmac.test.ts (hunk 2 above)
Comment (ohitslaurence) [reply src/lib/hmac.test.ts:304]: Fixed — replaced with golden test vectors computed from the platform's Node crypto implementation. Three pinned signatures now assert exact hex digests for POST with body, GET without body, and GET with canonicalized query params.

File: src/lib/client.ts (hunk 6 above)
Comment (ohitslaurence) [reply src/lib/client.ts:180]: Fixed — SpritzApiClient constructor now throws if integratorSecret is provided without integrationKey.

File: src/lib/hmac.ts (hunk 5 above)
Comment (ohitslaurence) [reply src/lib/hmac.ts:42]: Skipping — the platform server uses the exact same keys() + get() pattern in its canonicalizeQueryString. Changing only the SDK side would break signature compatibility on duplicate keys. Both need to be fixed together if at all, and the REST API doesn't use duplicate query param keys.

File: src/lib/client.ts (hunk 1 above)
Comment (ohitslaurence) [reply src/lib/client.ts:65]: Skipping — inherited pattern from integrationKey and apiKey. Fixing just the new param would be inconsistent; fixing all three is a separate cleanup.
